File: src/pathAnalysis.js
/**
 * Path Analysis Functions
 * 
 * Pure computational functions for analyzing paths in a graph snapshot.
 * These functions have NO Neo4j dependency - they work on in-memory data structures.
 */

const config = require('./config');

/**
 * @typedef {import('./providers/GraphDataProvider').GraphSnapshot} GraphSnapshot
 */

/**
 * Find top N paths by traffic volume (bottleneck throughput)
 * Uses min(edge.rate) along path as proxy for path throughput
 * Hard-capped to prevent combinatorial explosion
 * 
 * @param {GraphSnapshot} snapshot - Graph snapshot
 * @param {string} targetServiceId - Target service ID
 * @param {number} maxDepth - Maximum hops (edges) in path
 * @param {number} [maxPaths] - Maximum paths to return (default from config)
 * @returns {Array<{path: string[], pathRps: number}>}
 */
function findTopPathsToTarget(snapshot, targetServiceId, maxDepth, maxPaths = config.simulation.maxPathsReturned) {
    const paths = [];
    const visited = new Set();
    
    // DFS to enumerate paths (limited by maxPaths hard cap)
    // Uses hop-based depth: hops = currentPath.length - 1 (edges, not nodes)
    function dfs(currentId, currentPath, minRate) {
        if (paths.length >= maxPaths * 2) return; // Safety: early exit at 2x limit
        
        const hops = currentPath.length - 1; // hops = number of edges traversed
        
        // Found target with at least 1 hop
        if (currentId === targetServiceId && hops >= 1) {
            paths.push({
                path: [...currentPath],
                pathRps: minRate
            });
            return;
        }
        
        // Stop exploring if we've reached max hops
        if (hops >= maxDepth) return;
        
        // Sort outgoing edges for determinism: by rate desc, then target name asc
        const outgoing = (snapshot.outgoingEdges.get(currentId) || [])
            .slice()
            .sort((e1, e2) => (e2.rate - e1.rate) || e1.target.localeCompare(e2.target));
        
        for (const edge of outgoing) {
            if (visited.has(edge.target)) continue; // Prevent cycles
            
            visited.add(edge.target);
            currentPath.push(edge.target);
            
            const newMinRate = Math.min(minRate, edge.rate);
            dfs(edge.target, currentPath, newMinRate);
            
            currentPath.pop();
            visited.delete(edge.target);
        }
    }
    
    // Start DFS from all nodes (except target), sorted for determinism
    const startNodeIds = Array.from(snapshot.nodes.keys()).sort((a, b) => a.localeCompare(b));
    for (const nodeId of startNodeIds) {
        if (nodeId === targetServiceId) continue;
        if (paths.length >= maxPaths * 2) break;
        
        visited.clear();
        visited.add(nodeId);
        dfs(nodeId, [nodeId], Infinity);
    }
    
    // Sort by pathRps descending (already deterministic via sorted exploration)
    paths.sort((a, b) => b.pathRps - a.pathRps);
    return paths.slice(0, maxPaths);
}

module.exports = {
    findTopPathsToTarget
};

File: pkg/simulation/path_analysis.go
package simulation

import (
	"math"
	"sort"
)

func FindTopPathsToTarget(snapshot *GraphSnapshot, targetServiceId string, maxDepth int, maxPaths int) []BrokenPath {
	var paths []BrokenPath
	visited := make(map[string]bool)

	var startNodeIds []string
	for k := range snapshot.Nodes {
		startNodeIds = append(startNodeIds, k)
	}
	sort.Strings(startNodeIds)

	var dfs func(currentId string, currentPath []string, minRate float64)
	dfs = func(currentId string, currentPath []string, minRate float64) {
		if len(paths) >= maxPaths*2 {
			return
		}

		hops := len(currentPath) - 1

		if currentId == targetServiceId && hops >= 1 {

			pathCopy := make([]string, len(currentPath))
			copy(pathCopy, currentPath)
			paths = append(paths, BrokenPath{
				Path:    pathCopy,
				PathRps: minRate,
			})
			return
		}

		if hops >= maxDepth {
			return
		}

		edges := snapshot.OutgoingEdges[currentId]

		sort.Slice(edges, func(i, j int) bool {
			if edges[i].Rate != edges[j].Rate {
				return edges[i].Rate > edges[j].Rate
			}
			return edges[i].Target < edges[j].Target
		})

		for _, edge := range edges {
			if visited[edge.Target] {
				continue
			}

			visited[edge.Target] = true
			newPath := append(currentPath, edge.Target)
			newMinRate := math.Min(minRate, edge.Rate)

			dfs(edge.Target, newPath, newMinRate)

			delete(visited, edge.Target)
		}
	}

	for _, nodeId := range startNodeIds {
		if nodeId == targetServiceId {
			continue
		}
		if len(paths) >= maxPaths*2 {
			break
		}

		for k := range visited {
			delete(visited, k)
		}
		visited[nodeId] = true

		dfs(nodeId, []string{nodeId}, math.Inf(1))
	}

	sort.Slice(paths, func(i, j int) bool {
		return paths[i].PathRps > paths[j].PathRps
	})

	if len(paths) > maxPaths {
		return paths[:maxPaths]
	}
	return paths
}
